File: J9Ch/src/S_4_Binary_Search_Sweep_Line/Required_10/_141_Sqrtx_Easy.java
package S_4_Binary_Search_Sweep_Line.Required_10;

/**
141
Sqrt(x)

 * Created by tianhuizhu on 6/28/17.
 */
public class _141_Sqrtx_Easy {
    class Solution {
        /**
         * @param x: An integer
         * @return: The sqrt of x
         */
        public int sqrt(int x) {
            // find the last number which square of it <= x
            long start = 1, end = x;
            while (start + 1 < end) {
                long mid = start + (end - start) / 2;
                if (mid * mid <= x) {
                    start = mid;
                } else {
                    end = mid;
                }
            }

            if (end * end <= x) {
                return (int) end;
            }
            return (int) start;
        }
    }
}

File: J9Ch/src/J_2_Binary_Search/Related_8/python_cpp/_141_Sqrtx_Easy.cpp
/**
141
Sqrt(x)
*/

class Solution {
public:
    /**
     * @param x: An integer
     * @return: The sqrt of x
     */
    int sqrt(int x) {
        // write your code here
        long left = 0;
        if (x == 1)
            return 1;
        long right = x;
        long mid = left + (right-left)/2;
        while (left+1 < right) {
            if (x > mid*mid) {
                left = mid;
            } else if (x < mid*mid) {
                right = mid;
            } else {
                return mid;
            }
            mid = left + (right-left)/2;
        }
        return left;
    }
};